### ISSUE TITLE
SoftmaxFocalLoss with weights

### ISSUE THREAD

**Issue by ustclbh** [cfzd/Ultra-Fast-Lane-Detection#95]
Hi,
I am curious about the form of weights when we calculate Loss term，**L<sub>cls</sub>**, using the forward function in class SoftmaxFocalLoss (defined at utils/loss.py, line 26). See the snippet below,
```
class SoftmaxFocalLoss(nn.Module):
    def __init__(self, gamma, ignore_lb=255, *args, **kwargs):
        super(SoftmaxFocalLoss, self).__init__()
        self.gamma = gamma
        self.nll = nn.NLLLoss(ignore_index=ignore_lb)

    def forward(self, logits, labels):
        scores = F.softmax(logits, dim=1)
        factor = torch.pow(1.-scores, self.gamma)
        log_score = F.log_softmax(logits, dim=1)
       log_score = factor * log_score
        loss = self.nll(log_score, labels)
        return loss
```
Here is the question:
```
factor = torch.pow(1.-scores, self.gamma)
log_score = factor * log_score
loss = self.nll(log_score, labels)
```
These three lines describes the loss with the form **loss = -log((1-Prob)^2 * Prob)**, which is shown below.
![D( 1BD@N(5ML}@LG73@28%L](https://user-images.githubusercontent.com/31311295/93783887-d80bd180-fc5e-11ea-8ebd-560a3e212925.png)
As shown in the fig, it seems current loss is pushing the probility to somewhere closing to 0.3, instead of 1, when we optimize the loss. Isn't it weird?

Should it more reasonable to replace the current weighted loss with the vanilla CrossEntropy loss, i.e., **loss = -log(Prob)**. So we get smaller loss when the predicted Prob get close to 1.
![2 7T JRTJKF1R@CSWR~VPD2](https://user-images.githubusercontent.com/31311295/93783897-db9f5880-fc5e-11ea-90ae-a04a0f17df5e.png)

**Comment by cfzd** [comment]
@ustclbh 
I think it should be correct. We use two variables, which are `logit` and `prob` to, indicate the vanilla output of the network and the normalized probability.
```
logit = network(x)
prob = softmax(logit)
```
For cross-entropy loss, the loss calculation could be:
```
log_porb = log(prob)
loss = nll_loss(log_prob)
```
The mathematical definition is `loss = nll(log(prob))`.

For the focal loss, the calculation is:
```
scores = F.softmax(logits, dim=1)              ===> prob             
factor = torch.pow(1.-scores, self.gamma)      ===> (1-prob)^gamma
log_score = F.log_softmax(logits, dim=1)       ===> log(prob)
log_score = factor * log_score                 ===> (1-prob)^gamma * log(prob)
loss = self.nll(log_score, labels)             ===> loss = nll((1-prob)^gamma * log(prob))
```
So the `log(prob)` of CE corresponds to the `(1-prob)^gamma * log(prob)` of focal loss.

The image of `(1-prob)^gamma * log(prob)` can be found here:
https://github.com/clcarwin/focal_loss_pytorch/blob/master/images/focal_loss.png

**Comment by ustclbh** [comment]
Yeah, you are right.  I got a misunderstanding of NLLloss. Thanks for your explanation!
